target/bin/updatemailuser

```diff
@@ -10,25 +10,96 @@
 
 DATABASE=${DATABASE:-/tmp/docker-mailserver/postfix-accounts.cf}
 
-USER="${1}"
+MAIL_CRYPT_ENABLED=false
+
+function __usage
+{
+  echo -e "\e[35mUPDATEMAILUSER\e[31m(\e[93m8\e[31m)
+
+\e[38;5;214mNAME\e[39m
+    updatemailuser - allows user to update an existing email user
+      If mail_crypt is enabled, you'll need to provide the email user's old password.
+
+\e[38;5;214mSYNOPSIS\e[39m
+    updatemailuser <EMAIL ADDRESS> [ OPTIONS ] [<PASSWORD>]
+
+\e[38;5;214mOPTIONS\e[39m
+    \e[94mGeneric Program Information\e[39m
+        help       Print the usage information.
+
+\e[38;5;214mEXIT STATUS\e[39m
+    Exit status is 0 if command was successful, and 1 if there was an error.
+"
+}
+
+if [[ ${1} == 'help' ]]
+then
+  __usage
+  exit 0
+fi
+
+# Check if mail_crypt/per user storage encryption is enabled
+if ! grep '^#.*auth-passwdfile-mailcrypt' /etc/dovecot/conf.d/10-auth.conf &>/dev/null
+then
+  MAIL_CRYPT_ENABLED=true
+fi
+
+while getopts "hp:" OPT
+do
+  case ${OPT} in
+    p )
+      CRYPTOKEY_UPDATE_PASSWORD="${OPTARG}"
+      ;;
+    h )
+      __usage
+      exit 0
+      ;;
+    * )
+      __usage
+      echo "Invalid option: -${OPT}" >&2
+      ;;
+  esac
+done
+shift $(( OPTIND - 1 ))
+
+
+FULL_EMAIL="${1}"
 shift
 PASSWD="${*}"
 
-function usage  {	echo "Usage: updatemailuser <user@domain.tld> [password]" ; }
+[[ -z ${FULL_EMAIL} ]] && { usage ; errex "no username specified" ; }
 
-[[ -z ${USER} ]] && { usage ; errex "no username specified" ; }
+USER="${FULL_EMAIL%@*}"
+DOMAIN="${FULL_EMAIL#*@}"
 
 if [[ -z ${PASSWD} ]]
 then
-  read -r -s -p "Enter Password: " PASSWD
+  read -r -s -p "Enter New Password: " PASSWD
   echo
-  [[ -z ${PASSWD} ]] && errex "Password must not be empty"
+  [[ -z ${PASSWD} ]] && errex "password must not be empty"
 fi
 
-HASH="$(doveadm pw -s SHA512-CRYPT -u "${USER}" -p "${PASSWD}")"
+HASH="$(doveadm pw -s SHA512-CRYPT -u "${FULL_EMAIL}" -p "${PASSWD}")"
 
 # Protect config file with lock to avoid race conditions
 touch "${DATABASE}"
 create_lock "$(basename "$0")"
-grep -qi "^$(escape "${USER}")|" "${DATABASE}" 2>/dev/null || errex "User \"${USER}\" does not exist"
-sed -i "s ^""${USER}""|.* ""${USER}""|""${HASH}"" " "${DATABASE}"
+grep -qi "^$(escape "${FULL_EMAIL}")|" "${DATABASE}" 2>/dev/null || errex "user \"${FULL_EMAIL}\" does not exist"
+sed -i "s ^""${FULL_EMAIL}""|.* ""${FULL_EMAIL}""|""${HASH}"" " "${DATABASE}"
```

Comment:
Might want to convert this to `sedfile` as well? I'm not sure what the expect text you're matching is, I removed what looks like unnecessary `"`, but also some whitespace which might be a mistake on my part.

Also doesn't the syntax usually start with `s|` or are you using blank space as the delimiter? I think you might have, so I collapsed whitespace and escaped the `|` that you're presumably matching on.

```suggestion
sedfile -i "s|^${FULL_EMAIL}\|.*|${FULL_EMAIL}\|${HASH}|" "${DATABASE}"
```

Reply:
@casperklein may be able to weigh in, but it looks like I made the right assumption.

`git blame` [shows it was implemented 4 years ago](https://github.com/docker-mailserver/docker-mailserver/pull/897/commits/054cf992dc129f69d974e3583e12e8c8bd5ff026#diff-3f8b407f9e4682b80a37fb4323651214d70fbd3dcd22568292d6e6d5e785f714R26). It lacks details as to why they used whitespace as the delimiter, below it they added a commented out version as well:

```sh
sed -i "s ^"$USER"|.* "$USER"|"$PASSWD" " $DATABASE
#sed -i "s/^"$(escape "$USER")"|*/"$USER"|"$(doveadm pw -s SHA512-CRYPT -u "$USER" -p "$PASSWD")"/" $DATABASE
```

The `escape` method is no longer run on these vars we insert though, not sure if that will cause problems with `sed`, eg prior to your PR, this was `${USER}` not `${FULL_EMAIL}` (which contains special chars like `.`) right? A `${HASH}` may also contain special characters that mess with this :man_shrugging: 

Note that the escape method looks fairly basic:

https://github.com/docker-mailserver/docker-mailserver/blob/00c8a7b07fe0607e0da512d236e1461428492e64/target/scripts/helper-functions.sh#L13-L16 

---

Should be fine to adopt the `|`, possible issues you're currently experiencing may need to investigate dynamic inputs to `sed`. I'm not sure if we have any restrictions on user input (user/account name or password) that could be malicious in these contexts without sanitizing.